core/commands/storage/upload/upload/dc_repair_router.go

```diff
@@ -202,44 +212,338 @@ returns the repairer's signed contract to the invoker.`,
+				}
+
+				_, err = downloadAndRebuildFile(ctxParams, repairContract.FileHash, repairContract.LostShardHash)
+				if err != nil {
+					return err
+				}
+
+				err = challengeShards(ctxParams, repairContract)
+				if err != nil {
+					return err
+				}
+
+				fileStatus, err := uploadShards(ctxParams, repairContract)
+				if err != nil {
+					return err
+				}
+
+				err = submitFileStatus(ctxParams.Ctx, ctxParams.Cfg, fileStatus)
 				if err != nil {
 					return err
 				}
-				return nil
-			})
-			if err != nil {
-				return err
 			}
 			return nil
 		}()
 		if err != nil {
-			log.Errorf("repairer signed and submitted contract error: %s", err.Error())
+			log.Errorf("repair error: %s", err.Error())
 		}
 	}()
 }
 
+func submitSignedRepairContract(ctxParams *uh.ContextParams, params *RepairContractParams) (*guardpb.RepairContractResponse, error) {
+
+	repairReq := &guardpb.RepairContract{
+		FileHash:             params.FileHash,
+		FileSize:             params.FileSize,
+		RepairPid:            params.RepairPid,
+		LostShardHash:        params.LostShardHashes,
+		RepairSignTime:       time.Now().UTC(),
+		RepairContractId:     params.RepairContractId,
+		DownloadContractId:   params.DownloadContractId,
+		RepairRewardAmount:   params.RepairRewardAmount,
+		DownloadRewardAmount: params.DownloadRewardAmount,
+	}
+	sig, err := crypto.Sign(ctxParams.N.PrivateKey, repairReq)
+	if err != nil {
+		return nil, err
+	}
+	repairReq.RepairSignature = sig
+	var repairResp *guardpb.RepairContractResponse
+	err = grpc.GuardClient(ctxParams.Cfg.Services.GuardDomain).WithContext(ctxParams.Ctx, func(ctx context.Context,
+		client guardpb.GuardServiceClient) error {
+		repairResp, err = client.SubmitRepairContract(ctx, repairReq)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return repairResp, nil
+}
+
+func uploadShards(ctxParams *uh.ContextParams, repairContract *guardpb.RepairContract) (*guardpb.FileStoreStatus, error) {
+	repairContractReq := &guardpb.RequestRepairContracts{
+		FileHash:       repairContract.FileHash,
+		RepairNode:     repairContract.RepairPid,
+		RepairSignTime: repairContract.RepairSignTime,
+	}
+	sig, err := crypto.Sign(ctxParams.N.PrivateKey, repairContractReq)
+	if err != nil {
+		return nil, err
+	}
+	repairContractReq.RepairSignature = sig
+	var statusMeta *guardpb.FileStoreStatus
+	err = grpc.GuardClient(ctxParams.Cfg.Services.GuardDomain).WithContext(ctxParams.Ctx, func(ctx context.Context,
+		client guardpb.GuardServiceClient) error {
+		repairContractResp, err := client.RequestForRepairContracts(ctx, repairContractReq)
+		if err != nil {
+			return err
+		}
+		statusMeta = repairContractResp.Status
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	contracts := statusMeta.Contracts
+	if len(contracts) <= 0 {
+		return nil, errors.New("length of contracts is 0")
+	}
+	ssId, _ := uh.SplitContractId(contracts[0].ContractId)
```

Comment:
cannot get ssId from exist contract as so far the contractId was construct in guard, which did not know the session information of repair node.

Reply:
Created a new session here.